numpy/core/src/umath/ufunc_object.c

```diff
@@ -3613,46 +3613,76 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
         return NULL;
     }
 
-    /* Get the identity */
-    identity = _get_identity(ufunc, &reorderable);
-    if (identity == NULL) {
+    /* Get the reduction dtype */
+    if (reduce_type_resolver(ufunc, arr, odtype, &dtype) < 0) {
+        Py_DECREF(initial);
         return NULL;
     }
 
+
+    PyArrayObject *initial_arr = NULL;
+
     /* Get the initial value */
     if (initial == NULL || initial == NoValue) {
-        initial = identity;
-
+        PyObject *identity;
         /*
         * The identity for a dynamic dtype like
         * object arrays can't be used in general
         */
-        if (initial != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
-            Py_DECREF(initial);
-            initial = Py_None;
-            Py_INCREF(initial);
+        if (PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
+            identity = Py_None;
+            Py_INCREF(identity);
+        }
+        else {
+            /* Get the identity */
+            identity = _get_identity(ufunc, &reorderable);
+            if (identity == NULL) {
+                return NULL;
+            }
+        }
```

Comment:
Accidentally changed the `reorderable` logic here, and didn't mean to. Was just trying to avoid asking for the identity if we didn't need it.